zooniverse/front-end-monorepo
# Subject Group Viewer: set optional max grid width/height

## PR Overview

Package: `lib-classifier`
Related PR: [Add ADR-32: Subject Viewer Layouts (aka subject viewer sizing/fitting behaviour)](https://github.com/zooniverse/front-end-monorepo/pull/2229)

This PR allows project owners to set an optional maximum grid width/height in the config.

- Context (the problem): due to the Subject Viewer's default behaviour of "fit to width", there are scenarios where tall or square grids will have their bottom half "cut off" in the visible browser space on _wide_ browser screens.
- Implementation (the solution): using the workflow config, allow project owners to limit the visible dimensions of their grids
  - since each project owner will have a different grid setup, it makes sense that they should be able to also decide how the grid should "best" look in terms of "fitting" into the available visible space.
  - This obviously won't work for EVERY project for EVERY browser (that would require dynamic calculations) but it's a good enough solution for most users.
- To enable the optional max grid width/height, the workflow configuration needs some additional fields:
  ```
  workflow.configuration = {
    subject_viewer_config: {
      grid_max_width: '1000px',
      grid_max_height: '50vh',
      ...
    },
    ...
  }
  ```
  - This example shows a limit for both width (1000 pixels) and height (50% viewport height). I mean, obviously, in practice you just limit one of the dimensions, and you pick either pixels or viewport height, you don't mix and match like a madman. 
  - If no `grid_max_width` or `grid_max_height` is set, the default "fit grid to available width" behaviour triggers.

### Testing

URL: https://local.zooniverse.org:8080/?env=staging&project=1900&workflow=3412 (Survos Testing, staging)

Testing steps:
- Apply (via workflow config) the appropriate max-width/max-height you want to test for.
- Open the testing project URL.
- Resize your browser window.
  - Expect that the grid should fit the available width BUT will be limited to the maximum max-width/max-height that you've set, whichever is lower.
- With various browser window sizes, attempt to zoom in, zoom out, pan, and make annotations.
  - Expect that all these actions should work as expected 
- Remove the max-width/max-height and test again.
  - Confirm that by default, the "fit grid to available width" behaviour triggers.

⚠️ As of Friday evening, I haven't edited the testing project to have the correct workflow configuration yet. (I'mma haveta do some PFE editings, yerp.) If you want to test this PR on its own, I'm afraid you need to open up `SubjectGroupViewerContainer` and change `const DEFAULT_GRID_MAX_WIDTH = ''` to `const DEFAULT_GRID_MAX_WIDTH = '200px'` or something.

⚠️ Please note that the max-width/max-height logic is implemented via CSS, so... keep an eye out for CSS shenanigans across browsers. (e.g. some browsers might now support `vw` and `vh`. I know Safari 12 has different logic for fitting SVGs via height, but I think I've for solved that at least.)

### Status

Ready for review? You can definitely review the code, but to make functional testing easier I _really_ should add the workflow config to the test project.
